pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+)
+
+const (
+	attributeReferencerTagName = "attribute_referencer"
+)
+
+// resourceNotReadyErr is a custom error used for references validation
+type resourceNotReadyErr struct{ msg string }
+
+func newResourceNotReadyErr(format string, args ...interface{}) *resourceNotReadyErr {
+	return &resourceNotReadyErr{fmt.Sprintf(format, args...)}
+}
+
+func (r *resourceNotReadyErr) Error() string {
+	return r.msg
+}
+
+// LocalObjectGetter is a function that is used to retrieve local object references
+type LocalObjectGetter func(ref corev1.LocalObjectReference, obj runtime.Object) error
+
+// AttributeReferencer is an interface for referencing and resolving
+// cross-resource attribute references. See
+// https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-cross-resource-referencing.md
+// for more information
+type AttributeReferencer interface {
+	// Build retrieves and builds the referenced attribute
+	Build(getter LocalObjectGetter) (value string, err error)
+
+	// Assign accepts a resource object, and assigns the given value to the corresponding property
+	Assign(mgd Managed, value string) error
```

Comment:
Nitpick: `Managed` parameters are named `mg`, not `mgd` almost everywhere else in the codebase.